dask/bytes/core.py

```diff
@@ -1,22 +1,152 @@
 from __future__ import print_function, division, absolute_import
 
 import io
+import os
 
 from toolz import merge
+from warnings import warn
 
 from .compression import seekable_files, files as compress_files
 from .utils import SeekableFile
-from ..compatibility import PY2
+from ..compatibility import PY2, unicode
 from ..delayed import delayed
-from ..utils import infer_storage_options, system_encoding
+from ..utils import (infer_storage_options, system_encoding,
+                     build_name_function, infer_compression)
 
 delayed = delayed(pure=True)
 
 # Global registration dictionaries for backend storage functions
 # See docstrings to functions below for more information
 _read_bytes = dict()
+_write_bytes = dict()
 _open_files = dict()
 _open_text_files = dict()
+_write_files = dict()
+
+
+def write_block_to_file(data, f, compression, encoding):
+    """
+    Parameters
+    ----------
+    data : data to write
+        Either str/bytes, or iterable producing those, or something file-like
+        which can be read.
+    f : file-like
+        backend-dependent file-like object
+    compression : string
+        a key of `compress_files`
+    encoding : string (None)
+        if a string (e.g., 'ascii', 'utf8'), implies text mode, otherwise no
+        encoding and binary mode.
+    """
+    original = False
+    f2 = f
+    f = SeekableFile(f)
+    if compression:
+        original = True
+        f = compress_files[compression](f, mode='wb')
```

Comment:
Would it be cleaner if we made a more tightly-scoped function `write_bytes_to_file(data, f, compression)` and then put the burden on the caller to correctly encode their text data if they have it?  This might help to avoid a lot of the complex in this function.